compiler/optimizer/demand.go

```diff
@@ -0,0 +1,177 @@
+package optimizer
+
+import (
+	"fmt"
+	"github.com/brimdata/zed/compiler/ast/dag"
+)
+
+type Demand interface {
+	isDemand()
+}
+
+// TODO Think about normal form for Demand.
+
+// A nil Demand means 'demand nothing'.
+type DemandAll struct{}
+type DemandKey struct {
+	Key   string
+	Value Demand // Not nil
+}
+type DemandUnion [2]Demand // Disjoint. Not nil.
+
+func (demand DemandAll) isDemand()   {}
+func (demand DemandKey) isDemand()   {}
+func (demand DemandUnion) isDemand() {}
+
+func demandKey(key string, value Demand) Demand {
+	if value == nil {
+		return nil
+	}
+	return DemandKey{
+		Key:   key,
+		Value: value,
+	}
+}
+
+func demandUnion(a Demand, b Demand) Demand {
+	if a == nil {
+		return b
+	}
+	if b == nil {
+		return a
+	}
+
+	if _, ok := a.(DemandAll); ok {
+		return a
+	}
+	if _, ok := b.(DemandAll); ok {
+		return b
+	}
+
+	if a, ok := a.(DemandKey); ok {
+		if b, ok := b.(DemandKey); ok {
+			if a.Key == b.Key {
+				return DemandKey{
+					Key:   a.Key,
+					Value: demandUnion(a.Value, b.Value),
+				}
+			}
+		}
+	}
+
+	return DemandUnion([2]Demand{a, b})
+}
+
+func demandForSeq(seq dag.Seq) map[*dag.Op]Demand {
+	demands := make(map[*dag.Op]Demand)
```

Comment:
There are no unique ids for ops and I didn't want to add a demand field to every op, so I'm using pointers as map keys.

Reply:
This seems reasonable but I think you want map[dag.Op] instead of map[*dag.Op]

Reply:
Hmmm. If I write `Seq([op, op])` then they both have the same value as `dag.Op` but different values as `*dag.Op`. But even `*dag.Op` is not safe if two different seqs share an underlying array.

Reply:
I'd love to just do `map[[]int]` but alas...

Reply:
Maybe we can number them in post-order or something.

Reply:
I don't think there is a satisfying solution without invasive changes, so I'll just use `dag.Op` as the key and settle for checking that there are no duplicate ops.